odoo/addons/base/models/ir_rule.py

```diff
@@ -141,21 +149,36 @@ def _compute_domain(self, model_name, mode="read"):
         # browse user and rules as SUPERUSER_ID to avoid access errors!
         eval_context = self._eval_context()
         user_groups = self.env.user.groups_id
-        global_domains = []                     # list of domains
-        group_domains = []                      # list of domains
-        for rule in rules.sudo():
-            # evaluate the domain for the current user
-            dom = safe_eval(rule.domain_force, eval_context) if rule.domain_force else []
-            dom = expression.normalize_domain(dom)
-            if not rule.groups:
-                global_domains.append(dom)
-            elif rule.groups & user_groups:
-                group_domains.append(dom)
-
-        # combine global domains and group domains
+        global_domains = []  # list of lists of domains grouped by global_grouping_criteria
+        group_domains = []  # list of domains
+        for criteria, grouped_rules in rules.sudo().groupby(
+            lambda r: str(r.global_grouping_criteria)
+        ):
+            criteria = False if criteria == "False" else criteria  # FIXME: Don't str() in the groupby() somehow!
+            grouped_global_domains = []
+            for rule in grouped_rules:
+                # evaluate the domain for the current user
+                dom = safe_eval(rule.domain_force, eval_context) if rule.domain_force else []
+                dom = expression.normalize_domain(dom)
+                if not rule.groups and criteria:
+                    grouped_global_domains.append(dom)
+                elif not rule.groups and criteria:
+                    global_domains.append(dom)
+                elif rule.groups & user_groups:
+                    group_domains.append(dom)
+            if criteria:
+                global_domains.append(grouped_global_domains)
+
+        # AND together outer global domains, OR ones grouped by global_grouping_criteria
+        global_domains_expression = expression.AND([expression.OR(x) for x in global_domains])
```

Comment:
This is going to create "OR" expressions with one element in them, probably better to do only "OR" when len(x) > 1.

Reply:
tried (the simplified version);
```python
        if group_domains:
            # As we OR group_domains together like the inner lists in global_domains,
            # we can simply append them to global_domains to perform one expression on them
            global_domains.append(group_domains)
        return expression.AND([expression.OR(x) if len(x) > 1 else x for x in global_domains])
```
not sure if i'm missing something but got a lot of issues with how odoo resolves this. leaving the len() check out works fine though

Reply:
issues as in lots of these:
```
Traceback (most recent call last):
  File "/home/palabaster/projects/udes-vagrant/code/udes14/addons/udes_stock_permissions/tests/test_picking_type_access.py", line 139, in test_picktype_create_access
    self.StockPickingA.create_picking(
  File "/home/palabaster/projects/udes-vagrant/code/udes14/addons/udes_stock/models/stock_picking.py", line 308, in create_picking
    pickings = Picking.create(picking_values)
  File "<decorator-gen-179>", line 2, in create
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/api.py", line 329, in _model_create_single
    return self.browse().concat(*(create(self, vals) for vals in arg))
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/api.py", line 329, in <genexpr>
    return self.browse().concat(*(create(self, vals) for vals in arg))
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/addons/stock_picking_batch/models/stock_picking.py", line 18, in create
    res = super().create(vals)
  File "<decorator-gen-174>", line 2, in create
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/api.py", line 326, in _model_create_single
    return create(self, arg)
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/addons/stock/models/stock_picking.py", line 667, in create
    res = super(Picking, self).create(vals)
  File "<decorator-gen-137>", line 2, in create
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/api.py", line 347, in _model_create_multi
    return create(self, [arg])
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/addons/mail/models/mail_thread.py", line 264, in create
    threads = super(MailThread, self).create(vals_list)
  File "<decorator-gen-65>", line 2, in create
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/addons/base/models/ir_fields.py", line 534, in create
    recs = super().create(vals_list)
  File "<decorator-gen-13>", line 2, in create
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/api.py", line 348, in _model_create_multi
    return create(self, arg)
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/models.py", line 3877, in create
    records = self._create(data_list)
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/models.py", line 4051, in _create
    records.check_access_rule('create')
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/models.py", line 3347, in check_access_rule
    invalid = self - self._filter_access_rules_python(operation)
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/models.py", line 3402, in _filter_access_rules_python
    return self.sudo().filtered_domain(dom or [])
  File "/home/palabaster/projects/udes-vagrant/code/udes14/odoo/odoo/models.py", line 5307, in filtered_domain
    (key, comparator, value) = d
ValueError: not enough values to unpack (expected 3, got 1)
```

Reply:
with the len check we get extra tuples added
```
# without len check
['&', ('company_id', 'in', [1]), '|', '|', ('picking_type_id', 'in', []), ('picking_type_id', 'in', [975]), ('picking_type_id', 'in', [])]
# with len check
['&', (('company_id', 'in', [1]),), '|', '|', ('picking_type_id', 'in', []), ('picking_type_id', 'in', [975]), ('picking_type_id', 'in', [])]
```

Reply:
Fair enough, was just a nit.